## FWAKit Example 1 - linear referencing of point sites

fwakit includes functions that locates point features on the stream network by giving points a `blue_line_key` and `downstream_route_measure`. ltree type `fwa_watershed_code` and `local_watershed_code` values are also assigned for quick upstream/downstream queries.

### 1 - FWA setup
Once fwakit is installed (and the PostgreSQL/PostGIS/GDAL requirements):

- set the `FWA_DB` environment variable
- download and prep the FWA database as outlined in the README.  

To keep things small for this example, we'll use the test data (SALM watershed group) rather than the entire province, and only use three layers (lakes, streams, watersheds):

In [2]:
%env FWA_DB=postgresql://postgres:postgres@localhost:5432/fwakit

!fwakit create_db            
!fwakit download --source_url https://www.hillcrestgeo.ca/outgoing/fwakit/ --dl_path test_data --files FWA_BC.gdb.zip,FWA_STREAM_NETWORKS_SP.gdb.zip,FWA_WATERSHEDS_POLY.gdb.zip
!fwakit load --dl_path test_data -l fwa_lakes_poly,fwa_stream_networks_sp,fwa_watershed_poly_sp
!fwakit clean

env: FWA_DB=postgresql://postgres:postgres@localhost:5432/fwakit
Loading FWA source data to PostgreSQL database
Loading fwa_lakes_poly
Loading fwa_stream_networks_sp by watershed group
SALM
fwa_lakes_poly: cleaning
fwa_lakes_poly: adding ltree types
fwa_lakes_poly: indexing
fwa_watershed_groups_poly: cleaning
fwa_watershed_groups_poly: indexing
fwa_stream_networks_sp: cleaning
fwa_stream_networks_sp: adding ltree types
fwa_stream_networks_sp: indexing
fwa_watersheds_poly_sp: cleaning
fwa_watersheds_poly_sp: adding ltree types
fwa_watersheds_poly_sp: indexing


### 2 - Get some points

Grab Hydrometric Station points from Environment Canada and load to postgres:


In [3]:
# download
!wget http://dd.weather.gc.ca/hydrometric/doc/hydrometric_StationList.csv

# clean the header
!sed '1s/.*/id,name,lat,lon,prov,timezone/' hydrometric_StationList.csv > hydrostn.csv

# load to postgres
!ogr2ogr \
  --config PG_USE_COPY YES \
  -s_srs EPSG:4326 \
  -t_srs EPSG:3005 \
  -f PostgreSQL 'PG:host=localhost user=postgres dbname=fwakit password=postgres' \
  -lco OVERWRITE=YES \
  -overwrite \
  -lco GEOMETRY_NAME=geom \
  -oo X_POSSIBLE_NAMES=lon* \
  -oo Y_POSSIBLE_NAMES=lat* \
  -oo KEEP_GEOM_COLUMNS=NO \
  -sql "SELECT * FROM hydrostn WHERE prov='BC'" \
  -nln hydrostn \
  hydrostn.csv

--2018-02-25 21:27:08--  http://dd.weather.gc.ca/hydrometric/doc/hydrometric_StationList.csv
Resolving dd.weather.gc.ca (dd.weather.gc.ca)... 205.189.10.47
Connecting to dd.weather.gc.ca (dd.weather.gc.ca)|205.189.10.47|:80... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK
  Date: Mon, 26 Feb 2018 05:27:09 GMT
  Server: Apache
  Last-Modified: Mon, 26 Feb 2018 05:20:53 GMT
  Accept-Ranges: bytes
  Content-Length: 156529
  Content-Type: text/csv
  Via: 1.1 dd.weather.gc.ca
  X-UA-Compatible: IE=Edge
  Keep-Alive: timeout=5, max=100
  Connection: Keep-Alive
Length: 156529 (153K) [text/csv]
Saving to: ‘hydrometric_StationList.csv’

hydrometric_Station 100%[===================>] 152.86K   229KB/s    in 0.7s    

2018-02-25 21:27:09 (229 KB/s) - ‘hydrometric_StationList.csv’ saved [156529/156529]



Now check that data is loaded and in the right place:

In [4]:
%matplotlib inline

import geopandas as gpd
import matplotlib.pyplot as plt
import mplleaflet
import fwakit as fwa

db = fwa.util.connect()
df = gpd.read_postgis("SELECT id, prov, ST_Transform(geom, 4326) as geom FROM hydrostn", db.engine)
ax = df.plot()
mplleaflet.display(tiles='esri_worldtopo')

### 3 - Match points to stream network

In [26]:
fwa.reference_points('public.hydrostn', 'id', 'public.hydrostn_streams', 100, db=db)                        

What kind of match are we getting? 
Join the result to the source points and map, plus look at the event table itself - how far was each point from a stream?

In [27]:
stngdf = gpd.read_postgis("SELECT a.id, ST_Transform(a.geom, 4326) as geom "
                          "FROM hydrostn a "
                          "INNER JOIN hydrostn_streams b ON a.id=b.id", db.engine)
ax = stngdf.plot()
mplleaflet.display(tiles='esri_worldtopo') 

In [28]:
# look at the event table itself
import pandas as pd

df = pd.read_sql('SELECT * FROM hydrostn_streams', db.engine)
df

id  linear_feature_id       wscode_ltree    localcode_ltree  \
0  08HD006          710442195         920.722273  920.722273.097248   
1  08HD007          710448113  920.722273.416851  920.722273.416851   
2  08HD007          710447588         920.722273  920.722273.410699   
3  08HD015          710455835         920.722273  920.722273.606660   

                                  fwa_watershed_code  \
0  920-722273-000000-000000-000000-000000-000000-...   
1  920-722273-416851-000000-000000-000000-000000-...   
2  920-722273-000000-000000-000000-000000-000000-...   
3  920-722273-000000-000000-000000-000000-000000-...   

                                local_watershed_code  blue_line_key  \
0  920-722273-097248-000000-000000-000000-000000-...      354152994   
1  920-722273-416851-000000-000000-000000-000000-...      354148454   
2  920-722273-410699-000000-000000-000000-000000-...      354152994   
3  920-722273-606660-000000-000000-000000-000000-...      354152994   

   downstream_route_measure  distance_to_stream  
0              10983.087494           63.308356  
1                 27.194760           82.331672  
2              36576.107628           37.614846  
3              54628.950443           10.529124

Note that for the three stations in our test area, we have four results... station id `08HD007` has been joined to two streams: `blue_line_key` `354148454` and `354152994`. Both streams are within the specified 100m threshold distance from this station (at 82.3 and 37.6 m respectively, see the `distance_to_stream` column above). While the closest stream may indeed be the correct match, this may not always be the case. To better match our points to the correct stream we can check the station name against the stream name.

### 4 - Improve match of points to stream network

First, with this small example (one station, two matches) it is easy to manually compare the station name to the stream name and the distance to stream:

In [29]:
sql = """ -- find stations with more than one matched stream
          WITH multi_matched AS 
          (
          SELECT id, count(*)
          FROM hydrostn_streams
          GROUP BY id
          HAVING count(*) > 1
          )
         -- get names and distance to stream for these stations
         SELECT 
          m.id, 
          p.name as station_name,
          s.gnis_name as stream_name,
          e.distance_to_stream
        FROM multi_matched m
        INNER JOIN hydrostn p ON m.id = p.id
        INNER JOIN hydrostn_streams e ON p.id = e.id
        INNER JOIN whse_basemapping.fwa_stream_networks_sp s
        ON e.linear_feature_id = s.linear_feature_id"""
df = pd.read_sql(sql, db.engine)
df


id                      station_name   stream_name  distance_to_stream
0  08HD007  SALMON RIVER ABOVE MEMEKAY RIVER  Salmon River           37.614846
1  08HD007  SALMON RIVER ABOVE MEMEKAY RIVER     Kay Creek           82.331672

Conveniently, the station is on Salmon River, and the Salmon River is the closest stream. But we can do better than matching on the closest stream, postgres supplies several methods for string matching. 
As we are only comparing one value (the station name) to at most 2 or 3 matches (the streams within the threshold distance), a string comparison may work well. Lets try the `pg_trgm` extension first (https://www.postgresql.org/docs/10/static/pgtrgm.html) - adding a single line to the previous query, is the stream name similar to the station name?

In [30]:
db.execute('CREATE EXTENSION IF NOT EXISTS pg_trgm')

sql = """WITH multi_matched AS 
          (
          SELECT id, count(*)
          FROM hydrostn_streams
          GROUP BY id
          HAVING count(*) > 1
          )
      
         SELECT 
          m.id, 
          p.name as station_name,
          s.gnis_name as stream_name,
          s.gnis_name %% p.name as trgm_comparison,
          e.distance_to_stream
        FROM multi_matched m
        INNER JOIN hydrostn p ON m.id = p.id
        INNER JOIN hydrostn_streams e ON p.id = e.id
        INNER JOIN whse_basemapping.fwa_stream_networks_sp s
        ON e.linear_feature_id = s.linear_feature_id"""
df = pd.read_sql(sql, db.engine)
df

id                      station_name   stream_name  trgm_comparison  \
0  08HD007  SALMON RIVER ABOVE MEMEKAY RIVER  Salmon River             True   
1  08HD007  SALMON RIVER ABOVE MEMEKAY RIVER     Kay Creek            False   

   distance_to_stream  
0           37.614846  
1           82.331672

This works well for the test station. We can create a new table that refines the matching of station to stream with this logic:

In [32]:
sql = """CREATE TABLE hydrostn_referenced AS      
         SELECT 
          a.*, 
          p.name as station_name,
          str.gnis_name as stream_name
        FROM hydrostn_streams a
        INNER JOIN hydrostn p ON a.id = p.id
        INNER JOIN whse_basemapping.fwa_stream_networks_sp str
        ON a.linear_feature_id = str.linear_feature_id
        WHERE str.gnis_name %% p.name
      """
db.execute(sql)
df = pd.read_sql('SELECT * FROM hydrostn_referenced', db.engine)
df

id  linear_feature_id wscode_ltree    localcode_ltree  \
0  08HD006          710442195   920.722273  920.722273.097248   
1  08HD007          710447588   920.722273  920.722273.410699   
2  08HD015          710455835   920.722273  920.722273.606660   

                                  fwa_watershed_code  \
0  920-722273-000000-000000-000000-000000-000000-...   
1  920-722273-000000-000000-000000-000000-000000-...   
2  920-722273-000000-000000-000000-000000-000000-...   

                                local_watershed_code  blue_line_key  \
0  920-722273-097248-000000-000000-000000-000000-...      354152994   
1  920-722273-410699-000000-000000-000000-000000-...      354152994   
2  920-722273-606660-000000-000000-000000-000000-...      354152994   

   downstream_route_measure  distance_to_stream  \
0              10983.087494           63.308356   
1              36576.107628           37.614846   
2              54628.950443           10.529124   

                                 station_name   stream_name  
0                   SALMON RIVER NEAR SAYWARD  Salmon River  
1            SALMON RIVER ABOVE MEMEKAY RIVER  Salmon River  
2  SALMON RIVER ABOVE CAMPBELL LAKE DIVERSION  Salmon River